### DISCUSSION TITLE
Questions

### DISCUSSION THREAD

**Discussion by sQur1** [robert-hh/Micropython-Editor#29] [Ideas]
First, my apologies if this is not the place for this kind of question, but I didn't know where else to ask.

I was wondering if this library could be the backbone for a custom Distraction-free Writing Device where something like a Pico would be used in low frequency paired with a keyboard and a small LCD screen? It would be a standalone write-on-the-go little machine, not dependent on any other computers, but perhaps write the text files on a MicroSD card.

Thank you

**Comment by robert-hh** [comment]
We had a similar discussion with PR #5, and this user implemented a set-up with LCD and a small keypad. That discussion lead to the separation of the core edit module and a front end. One of the front ends is pye_lcd.py, which this user  made for a combination with a ST7789 display and CircuitPython. I recall having made a front-end for a SSD1963 display, but cannot find it: Anyhow, it was not difficult.
The problem is the input side. As far as I know, there is no support for a USB keyboard at the RP2040 port. But in any case, all of that would be more work with a less comfortable result than using a 50$ tablet with restricted software.

**Reply by sQur1** [reply]
Thank you for your input Robert. I'm looking to build one for very long battery life type of thing. Tablet is easy but very short battery life.

**Comment by robert-hh** [comment]
Most of the battery power is consumed by the display. So the display type is essential. A large LCD display consumes several 100mA for backlight. OLED devices consume power proportional to the number of enabled dots. So you need a passive display for low current consumption.

**Reply by sQur1** [reply]
That's good to know, thank you! I got a 20x4 monochrome LCD that I bought for this purpose. I might've been too optimistic in trying to build something like that from scratch with my limited amount of programming knowledge lol.

**Reply by robert-hh** [reply]
20x4 characters is not much space to write. I had assumed that you would uses a larger TFT display. But these character displays do not consume a lot of current, but it's still about 40mA for the backlight.

**Reply by robert-hh** [reply]
You could as well use viper code for fast access to the hardware. That's much faster than machine.mem. As short example below, which toggles an I/O pin. With mem32, I get ~140kHz, with viper 8 MHz.
```
from machine import Pin, mem32
gen = Pin(22, Pin.OUT)

def mem32toggle():
    pin = (1 << 22)
    while True:
        mem32[0x3FF44008] = pin
        mem32[0x3FF4400C] = pin

@micropython.viper
def vipertoggle():
    GPIO = ptr32(0x3FF44004)
    while True:
        GPIO[1] = 1 << 22
        GPIO[2] = 1 << 22
```

**Comment by DG12** [comment]
@sQur1  Is an inexpensive tablet from eBay an option with the ability to have 1 (or more) spare batteries if they could be swapped easily enough? 

Of course if you WANT to develop you own go for it! :-)
